SPEC.md

```diff
@@ -0,0 +1,334 @@
+
+  - `Nexus-Operation-State: succeeded`
+
+  **Body**: Arbitrary data conveying the operation's result. Headers should specify content type and encoding.
+
+- `204 No Content`: Operation completed successfully with an empty result or is still running.
+
+  **Headers**:
+
+  - `Nexus-Operation-State: succeeded | running`
+
+  **Body**: Empty.
+
+- `482 Operation Failed`: Operation completed as `failed` or `canceled`.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+  - `Nexus-Operation-State: failed | canceled`
+
+  **Body**: A JSON serialized [`Failure`](#failure) object.
+
+- `408 Request Timeout`: The specified `wait_deadline` expired prior to operation completion.
+
+  The caller may re-issue this request to start a new long poll.
+
+  **Body**: Unspecified.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+  **Body**: Unspecified.
+
+### Get Operation Info
+
+Retrieve operation details.
+
+#### Method
+
+`GET`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}`
+
+#### Response Codes
+
+- `200 OK`: Successfully retrieved info.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+  - `Nexus-Operation-State: failed | canceled`
+
+  **Body**:
+
+  A JSON serialized [`OperationInfo`](#operationinfo) object.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+## General Information on HTTP Response Codes
+
+The Nexus protocol follows standard HTTP practices:
+
+- `4xx` codes signify non-retryable framework-level errors.
+- `5xx` codes signify retryable framework-level errors.
+
+## Callback URLs
+
+Any HTTP URL can be used to deliver operation completions.
+
+Callers should ensure URLs contain sufficient information to correlate completions with initiators.
+
+For invoking a callback URL:
+
+- Include the `Nexus-Operation-State` header.
+- If state is `succeeded`, deliver non-empty results in the body with corresponding `Content-*` headers.
+- If state is `failed` or `canceled`, content type should be `application/json` and the body must have a serialized [`Failure`](#failure) object.
+- Upon successful completion delivery, the handler should reply with a `200 OK` status and an empty body.
+
+### Security
+
+There's no enforced security for callback URLs.
+When starting an operation, callers may embed a signed token into the URL, which can be verified upon delivery of completion.
+
+[rfc3986-section-2.3]: https://datatracker.ietf.org/doc/html/rfc3986#section-2.3
+
+### Q/A
+
+1. What is the purpose of the `482 Operation Failed` response code? This is not a standard HTTP response code.
+
+In internal discussions, we determined that there's value in having a specific status code for denoting operation
+failures.
+We chose a 4xx code because it is not meant to be retried and can be used by standard HTTP libraries and tools to
+easier distinguish failure vs. success.
+Non of the standard HTTP codes fit and we decided to define a custom status code.
```

Comment:
To sum up the thoughts on retry, we are basically saying "you can write your Nexus client to automatically retry any 5xx errors". This is a Nexus rule we're making to help client authors, not necessarily rooted in general HTTP spec anywhere.

If it's a 4xx, whether you want to retry is a logic question, we're just saying 5xx is safe for implicit-auto-retry logic.

Reply:
Since we're including this explanatory Q/A section, it might be worth explicitly saying something like

> Technically this is an abuse of 4xx, because an operation-level failure (for example because a required 3rd-party service was down) is not a client error. An alternative would be a 2xx, since the handler successfully drove the requested operation to a valid terminal state. However, we want to avoid requiring callers to inspect headers to determine whether a 2xx response contains their desired response or a `Failure` object.